[Clang][Comments] Add argument parsing for @throw @throws @exception

hdoc opened the pull request:
Doxygen allows for the `@throw`, `@throws`, and `@exception` commands to have an attached argument indicating the type being thrown. Currently, Clang's AST parsing doesn't support parsing out this argument from doc comments. The result is missing compatibility with Doxygen.

We would find it helpful if the AST exposed these thrown types as BlockCommandComment arguments so that we could generate better documentation.

This PR implements parsing of arguments for the `@throw`, `@throws`, and `@exception` commands. Each command can only have one argument, matching the semantics of Doxygen. We have also added unit tests to validate the functionality.

hdoc committed:
Comment parsing: add argument parsing for @throw @throws @exception

Doxygen allows for the @throw, @throws, and @exception commands to
have an attached argument indicating the type being thrown. Currently,
Clang's AST parsing doesn't support parsing out this argument from doc
comments. The result is missing compatibility with Doxygen.

We would find it helpful if the AST exposed these thrown types as
BlockCommandComment arguments so that we could generate better
documentation.

This PR implements parsing of arguments for the @throw, @throws, and
@exception commands. Each command can only have one argument, matching
the semantics of Doxygen. We have also added unit tests to validate
the functionality.

sdkrystian commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -149,6 +191,76 @@ class TextTokenRetokenizer {
+    SmallString<32> WordText;
+    const char *WordBegin = Pos.BufferPtr;
+    SourceLocation Loc = getSourceLocation();
+    StringRef ConstVal = StringRef("const");
```
Shouldn't we support `volatile` as well?

hdoc replied:
Yes, it should. We'll update the PR to support that. Thank you for noticing and mentioning this.

Likewise in the future we'll probably want to add support for `restrict`. It's not necessary for this PR since `throw/throws/exception` is only useful in C++ but when parsing Doxygen comments in C codebases we'll want to have the option of supporting that.

hdoc committed:
Fix comment to XML tests

hdoc committed:
Update XML output routine for Throws commands to fix XML validation

XML validation was failing due to the Exception XML being empty, since
the actual exception type was being parsed as an argument instead of
as a ParagraphComment.

This was a result of the change we made to argument parsing.

As a result, I updated the XML output to still output the argument
text in the Para XML, as it was emitted before.

hdoc committed:
Support more qualifiers, such as volatile

hdoc committed:
Refactor integer argument parsing and add corresponding test cases

hdoc commented:
Ping

sdkrystian commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -75,6 +75,25 @@ class TextTokenRetokenizer {
+  char peekNext(unsigned offset) const {
+    assert(!isEnd());
+    assert(Pos.BufferPtr != Pos.BufferEnd);
+    if (Pos.BufferPtr + offset <= Pos.BufferEnd) {
```
Shouldn't this be `if (Pos.BufferPtr + offset < Pos.BufferEnd)` (unless this depends of the buffer being null terminated)? Otherwise this allows for an `offset` which results in `Pos.BufferEnd` being read.

hdoc replied:
Good catch, fixed.

sdkrystian commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -89,6 +108,114 @@ class TextTokenRetokenizer {
   }
 
+  bool continueInt(SmallString<32> &NextToken) {
+    return NextToken.ends_with(StringRef("char")) ||
```
Don't really understand what this is for...

hdoc replied:
This is used during integral lexing to decide whether we should continue lexing the presumed integral type or not. The name was confusing, so I've renamed it from `continueInt` to `shouldContinueLexingIntegralType`

sdkrystian commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -149,6 +276,93 @@ class TextTokenRetokenizer {
+    const char *WordBegin = Pos.BufferPtr;
+    SourceLocation Loc = getSourceLocation();
+    StringRef ConstVal = StringRef("const");
+    StringRef PointerVal = StringRef("*");
```
What about pointers to members? Arrays? Function? More complex declarators like `void(*)(int, int)`?

hdoc replied:
For now, this doesn't support those values, and due to the complexity of implementing them and the rarity of their use I think it'd make more sense to add them in a follow-up PR. Let me know if this sounds reasonable

sdkrystian commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -149,6 +276,93 @@ class TextTokenRetokenizer {
+    StringRef ReferenceVal = StringRef("&");
+    bool ConstPointer = false;
+
+    while (!isEnd()) {
```
Some comments wouldn't hurt...

hdoc replied:
Added as part of the latest commit, along with some light refactoring to make it more readable. No doubt it is a complex piece of code

sdkrystian commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -149,6 +276,93 @@ class TextTokenRetokenizer {
+          if (!lexTemplate(WordText))
+            return false;
+        } else {
+          WordText.push_back(C);
```
Does this treat any non-whitespace character (other than `<`) as being part of a name?

hdoc replied:
As the code stands now, yes

cor3ntin replied:
Which I think is fine, we are just trying to find a boundary

hdoc committed:
Address review feedback

- Refactored some functionality with the help of clang-tidy
- Added comments to the complicated lexing functions
- Renamed variables to follow LLVM/Clang conventions

hdoc commented:
Ping

cor3ntin commented:
Before I do an in-depth review i think supporting arguments to `@throw` etc makes a lot of sense.
However, I don't think we want to have a `type-id` parser in the  comment parser. The grammar is complicated, and the approach seems brittle and hard to maintain.

Given use cases, I think it's sufficient to consider anything before a space as part of a type.
You can try to support things like `foo<const bar>` just by balancing `<>`,  but I really don't think we want to parse arbitrary types (including function types, arrays, lambdas, etc) as types that you can throw.

If people want to throw "unsigned int" exceptions, well... it won't work. I think it's a reasonable tradeoff.
99% of exceptions are going to be class types.

In the long term, if we really want to support arbitrary constructs there, I think it would take a more robust approach such as the one that was [proposed here (and subsequently abandoned) ](https://discourse.llvm.org/t/rfc-a-c-pseudo-parser-for-tooling/59217 
)

hdoc committed:
Parse only the space-delimited word after @throw

This is in response to a request by upstream to simplify the parser
previously developed.

hdoc commented:
@cor3ntin we appreciate the feedback. Likewise, we agree that the home-grown parser just for parsing `type-id`s in comments is less than ideal. We had some alternative approaches in mind, like using an instance of `clang::Lexer`, but went with our original approach at first.

I've refactored the PR and reduced the functionality to the subset you requested. I think that makes more sense from a maintainability perspective, as you highlighted. Please review the PR when you get a chance. Thank you.

cor3ntin commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -89,6 +89,31 @@ class TextTokenRetokenizer {
+  bool lexTemplate(SmallString<32> &WordText) {
+    unsigned IncrementCounter = 0;
+    while (!isEnd()) {
+      const char C = peek();
+      WordText.push_back(C);
+      consumeChar();
```
It would be nice to modify `consumeChar` to return the consumed char

hdoc replied:
I would prefer to do that as a separate refactor PR, I think it's outside the scope of this change.

cor3ntin commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -89,6 +89,31 @@ class TextTokenRetokenizer {
 
+  /// Extract a template type
+  bool lexTemplate(SmallString<32> &WordText) {
+    unsigned IncrementCounter = 0;
```
`BracketCount` would be a better name

hdoc replied:
That's a much better name, fixed and thank you.

cor3ntin commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -149,6 +174,54 @@ class TextTokenRetokenizer {
+      return false;
+    }
+
+    char *TextPtr = Allocator.Allocate<char>(Length + 1);
+
+    memcpy(TextPtr, WordText.c_str(), Length + 1);
+    StringRef Text = StringRef(TextPtr, Length);
+
+    formTokenWithChars(Tok, Loc, WordBegin, Length, Text);
```
I'd like to see that extracted in a separate function (and replace the few places where we do the exact same thing)

hdoc replied:
As above, I would prefer to do that as a separate refactor PR, I think it's outside the scope of this change.

cor3ntin commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -295,6 +367,7 @@ Parser::parseCommandArgs(TextTokenRetokenizer &Retokenizer, unsigned NumArgs) {
       Comment::Argument[NumArgs];
   unsigned ParsedArgs = 0;
   Token Arg;
+
```
Can you revert whitespace only changes?

hdoc replied:
Fixed, I try to keep the clang-format changes restricted to only the regions I touch but it looks like some sneaked through

cor3ntin reviewed (commented):
The general direction looks good now. I left a few nitpicky comments.

Thanks!

hdoc committed:
Rename IncrementCounter -> BracketCount

hdoc committed:
Fix comment string

hdoc committed:
Revert clang-format changes that sneaked through

hdoc commented on clang/include/clang/AST/CommentParser.h:
```diff
@@ -100,6 +100,11 @@ class Parser {
   ArrayRef<Comment::Argument>
   parseCommandArgs(TextTokenRetokenizer &Retokenizer, unsigned NumArgs);
 
+  /// Parse arguments for \\throws command supported args are in form of class
```
Fixed

hdoc committed:
Add some more tests to better specify edge case behavior

hdoc commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -89,6 +89,31 @@ class TextTokenRetokenizer {
   }
 
+  /// Extract a template type
+  bool lexTemplate(SmallString<32> &WordText) {
```
Added some more test cases that check this edge case behavior, like unequal numbers of opening and closing brackets as well as some other unusual edge cases.

cor3ntin reviewed (approved):
LGTM, Thanks
Will you need us to merge that for you?

hdoc commented:
@cor3ntin yes, please merge this as we do not have commit access.
